File: code/python/data_structures_article-12.py
"""Ring detection and topological sorting algorithm"""

from __future__ import annotations
from typing import Any


def solve(*args: Any, **kwargs: Any) -> Any:
    """Run this lesson exercise with positional inputs."""
    return _Solution().canFinish(*args, **kwargs)


class _Solution:
    def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:
        onPath = [False] * numCourses
        visited = [False] * numCourses
        hasCycle = [False]
        graph = self.buildGraph(numCourses, prerequisites)
        def traverse(s):
            if onPath[s]:
                hasCycle[0] = True
                """
                """
            if visited[s] or hasCycle[0]:
                return
            visited[s] = True
            onPath[s] = True
            for t in graph[s]:
                traverse(t)
            onPath[s] = False
        for i in range(numCourses):
            traverse(i)
        return not hasCycle[0]
    def buildGraph(self, numCourses, prerequisites):
        graph = [[] for _ in range(numCourses)]
        for from_, to in prerequisites:
            graph[from_].append(to)
        return graph

File: code/javascript/data_structures_article-12.js
/**
 * Ring detection and topological sorting algorithm
 */
export function solve(...args) {
  // Entry point for the course exercise: pass inputs directly.
  return traverse(...args);
}

var canFinish = function(numCourses, prerequisites) {
    const onPath = new Array(numCourses).fill(false);
    const visited = new Array(numCourses).fill(false);
    let hasCycle = false;
    const graph = buildGraph(numCourses, prerequisites);
    for (let i = 0; i < numCourses; i++) {
        traverse(graph, i);
    }
    return !hasCycle;
    function traverse(graph, s) {
        if (onPath[s]) {
            hasCycle = true;
            return;
        }
        if (visited[s] || hasCycle) {
            return;
        }
        visited[s] = true;
        onPath[s] = true;
        for (let t of graph[s]) {
            traverse(graph, t);
        }
        onPath[s] = false;
    }
    function buildGraph(numCourses, prerequisites) {
        const graph = new Array(numCourses).map(() => []);
        for (let edge of prerequisites) {
            const from = edge[1];
            const to = edge[0];
            graph[from].push(to);
        }
        return graph;
    }
};
